tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/RecordingZ3SolverContext.scala

```diff
@@ -0,0 +1,207 @@
+package at.forsyte.apalache.tla.bmcmt.smt
+
+import at.forsyte.apalache.tla.bmcmt.profiler.SmtListener
+import at.forsyte.apalache.tla.bmcmt.{Arena, ArenaCell}
+import at.forsyte.apalache.tla.lir.TlaEx
+import com.typesafe.scalalogging.LazyLogging
+
+
+object RecordingZ3SolverContext {
+  def apply(parentLog: Option[SmtLog], config: SolverConfig): RecordingZ3SolverContext = {
+    val context = new RecordingZ3SolverContext(parentLog, config)
+    parentLog match {
+      case Some(log) =>
+        log.replay(context.solver)
+
+      case None => ()
+    }
+    context
+  }
+}
+
+/**
+  * A wrapper around SolverContext that records SMT operations. The log is organized hierarchically,
+  * so a differential diff may be taken.
+  *
+  * @param parentLog the log of the parent context
+  * @param config solver configuration
+  *
+  * @author Igor Konnov
+  */
+@SerialVersionUID(700L)
+class RecordingZ3SolverContext private (parentLog: Option[SmtLog], val config: SolverConfig)
+    extends SolverContext with Serializable with LazyLogging {
+
+  import SmtLog._
+
+  private val solver = new Z3SolverContext(config)
+
+  /**
+    * The sequence of logs (the last added element is in the head),
+    * one per context, except the last log, which is maintained in lastLog.
+    * Every list in logStackRev is stored in the reverse order.
+    */
+  private var logStackRev: List[List[Record]] = List()
+
+  /**
+    * The current log, the last added element is in the head.
+    */
+  private var lastLogRev: List[Record] = List()
+
+  /**
+    * Extract a differential SMT log from the current context.
+    * @return SMT log
+    */
+  def extractLog(): SmtLog = {
+    val newRecords = (lastLogRev ++ logStackRev.flatten).reverse
+    new SmtLog(parentLog, newRecords)
+  }
+
+  /**
+    * Get the current context level, that is the difference between the number of pushes and pops made so far.
+    *
+    * @return the current level, always non-negative.
+    */
+  override def contextLevel: Int = solver.contextLevel
+
+  /**
+    * Save the current context and push it on the stack for a later recovery with pop.
+    */
+  override def push(): Unit = {
+    solver.push()
+    logStackRev = lastLogRev :: logStackRev
+    lastLogRev = List()
+  }
+
+  /**
+    * Pop the previously saved context. Importantly, pop may be called multiple times and thus it is not sufficient
+    * to save only the latest context.
+    */
+  override def pop(): Unit = {
+    solver.pop()
+    lastLogRev = logStackRev.head
+    logStackRev = logStackRev.tail
```

Comment:
1. Why not define this as `pop(1)`? 
2. Wont these crash when the stack is empty? Afaict, nothing ensures the logs won't be empty. Unless something guarantees this invariant, should either raise a clear exception when trying to pop a nonempty stack or (better, if not too inconvenient), make `pop(): Optional[Unit]` (same goes for `pop n`.

Reply:
Totally true. I will add an exception on the empty log. This would be an irrecoverable error though.